internal/registry/registry.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2021 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registry
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+
+	"github.com/tetratelabs/car/internal"
+	"github.com/tetratelabs/car/internal/httpclient"
+	"github.com/tetratelabs/car/internal/registry/docker"
+	"github.com/tetratelabs/car/internal/registry/github"
+)
+
+type registry struct {
+	host, path, baseURL string
+	roundTripper        http.RoundTripper
+}
+
+type contextRoundTripper struct {
+}
+
+func (f *contextRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return httpclient.TransportFromContext(req.Context()).RoundTrip(req)
+}
+
+// New returns a new instance of a remote registry
+func New(ref reference.Named) internal.Registry {
+	domain := reference.Domain(ref)
+	path := reference.Path(ref)
+	host := domain
+	var roundTripper http.RoundTripper
+	switch domain {
+	case "docker.io":
+		host = "index.docker.io"
+		roundTripper = docker.NewRoundTripper(path)
+	case "ghcr.io":
+		roundTripper = github.NewRoundTripper()
+	default:
+		roundTripper = &contextRoundTripper{}
+	}
+	baseURL := fmt.Sprintf("https://%s/v2/%s", host, path)
+	return &registry{host: host, path: path, baseURL: baseURL, roundTripper: roundTripper}
+}
+
+func (r *registry) String() string {
+	return r.baseURL
+}
+
+func (r *registry) GetImage(ctx context.Context, tag, platform string) (*internal.Image, error) {
+	client := r.httpClient(ctx)
+
+	// A tag can respond with either a multi-platform image or a single one, so we have to handle either.
+	images, err := r.getImageManifests(ctx, client, tag)
+	if err != nil {
+		return nil, err
+	}
+	if len(images) == 0 {
+		return nil, fmt.Errorf("image tag %s not found", tag)
+	}
+
+	// History (created_by for each layer) is not in the manifest, rather the config JSON.
+	configs, err := r.getImageConfigs(ctx, client, images)
+	if err != nil {
+		return nil, err
+	}
+
+	// Combine the two sources into the Image internal we need.
+	var result []*internal.Image
+	lastOSVersion := ""
+	for image := range images {
```

Comment:
I suggest to use `i` for indexes. This makes me feel like `images` would be a map and therefore `image` be a key of it.

Reply:
yeah that was a mistake I was trying to go back and un-go name everything per another comment of yours and made this mistake in the process :D

Reply:
should be a linter for that.. I would really not typically use more than one char for index. anyway good eye and done!